README.md
# Latest News and Versions
28 August 2023 - version 0.32 released
- Fixed bug in Bond OAS and ASW

24 August 2023 - Version 0.31 released

Schedule 
- Corrected bug in schedule generation
- Corrected bug in CDS protection leg integral

Many Bond Classes have been amended
- Changed FULL price to DIRTY price in functions UPDATE YOUR CODE PLEASE. APOLS for inconvenience.
- Removed face amount from bond class - how much you buy is not intrinsic to a bond
- Made number of ex-dividend days a member of bond class
- Added adjustment for ex-dividend dates to yield calculations
- Revised acccrued and principal functions to take face amount as input
- Updated document

29 May 2023 - Version 0.30 released
- Added PrettyPrint to required dependencies

22 Nov 22
Version 0.260 has been released and pushed to PyPI
- Create Date from python datetime
- Zero coupon bond class
- Fixed bug in bond payment date

31-Aug-2022
Version 0.240 has just been released and pushed to PyPI with changes
- Negative terms in date class
- Recovery rates do not default to standard value for CDS curves 

# DISCLAIMER

This software is distributed FREE AND WITHOUT ANY WARRANTY. 

Report any bugs or suggestions here as an issue. 

# CONTRIBUTORS WANTED !

If you have a knowledge of Quantitative Finance and a reasonable knowledge of Python, then please consider contributing to this project. There are small tasks and big tasks to be done. Just look in the list of Issues and you may find something you can do. Before you begin, please comment in the issue thread in case someone else may be working on that issue. Or you can contact me directly at dominic.okane at edhec.edu. 

If you are a user and require some additional functionality, then please add it as an issue.

# Quick Start Guide

FinancePy can be installed from pip using the following command:

`pip install financepy`

To upgrade an existing installation type:

`pip install --upgrade financepy`

I have encountered problems using Anaconda3-2020.07 due to some Numba and LLVMLite problems. However Anaconda3-2020.02 works.

## Using FinancePy in a Jupyter Notebook

Once financepy has been installed, it is easy to get started.

Just download the project and examine the set of Jupyter Notebooks in the notebooks folder.

A pdf manual describing all of the functions can be found in the project directory.

## Overview

FinancePy is a python-based library that is currently in beta version. It covers the following functionality:

* Valuation and risk models for a wide range of equity, FX, interest rate and credit derivatives.

Although it is written entirely in Python, it can achieve speeds comparable to C++ by using Numba. As a result the user has both the ability to examine the underlying code and the ability to perform pricing and risk at speeds which compare to a library written in C++.

The target audience for this library includes:

* Students of finance and students of python
* Academics teaching finance or conducting research into finance
* Traders wishing to price or risk-manage a derivative.
* Quantitative analysts seeking to price or reverse engineer a price.
* Risk managers wishing to replicate and understand price sensitivity.
* Portfolio managers wishing to check prices or calculate risk measures.
* Fund managers wanting to value a portfolio or examine a trading strategy.

Users should have a good, but not advanced, understanding of Python. In terms of Python, the style of the library has been determined subject to the following criteria:

1. To make the code as simple as possible so that those with a basic Python fluency can understand and check the code.
2. To keep all the code in Python so users can look through the code to the lowest level.
3. To offset the performance impact of (2) by leveraging Numba to make the code as fast as possible without resorting to Cython.
4. To make the design product-based rather than model-based so someone wanting to price a specific product can easily find that without having to worry too much about the model – just use the default – unless they want to. For most products, a Monte-Carlo implementation has been provided both as a reference for testing and as a way to better understand how the product functions in terms of payments, their timings and conditions.
5. To make the library as complete as possible so a user can find all their required finance-related functionality in one place. This is better for the user as they only have to learn one interface.
6. To avoid complex designs. Limited inheritance unless it allows for significant code reuse. Some code duplication is OK, at least temporarily.
7. To have good documentation and easy-to-follow examples.
8. To make it easy for interested parties to contribute.

In many cases the valuations should be close to if not identical to those produced by financial systems such as Bloomberg. However for some products, larger value differences may arise due to differences in date generation and interpolation schemes. Over time it is hoped to reduce the size of such differences.

Important Note:

* IF YOU HAVE ANY PRICING OR RISK EXAMPLES YOU WOULD LIKE REPLICATED, SEND SCREENSHOTS OF ALL THE UNDERLYING DATA, MODEL DETAILS AND VALUATION.
* IF THERE IS A PRODUCT YOU WOULD LIKE TO HAVE ADDED, SEND ME THE REQUEST.
* IF THERE IS FUNCTIONALITY YOU WOULD LIKE ADDED, SEND ME A REQUEST.

## The Library Design

The underlying Python library is split into a number of major modules:

* Utils - These are utility functions used to assist you with modelling a security. These include dates (Date), calendars, schedule generation, some finance-related mathematics functions and some helper functions.
* Market - These are modules that capture the market information used to value a security. These include interest rate and credit curves, volatility surfaces and prices.
* Models - These are the low-level models used to value derivative securities ranging from Black-Scholes to complex stochastic volatility models.
* Products - These are the actual securities and range from Government bonds to Bermudan swaptions.

Any product valuation is the result of the following data design:

**VALUATION** = **PRODUCT** + **MODEL** + **MARKET**

The interface to each product has a value() function that will take a model and market to produce a price.

## Author

Dominic O'Kane. I am a Professor of Finance at the EDHEC Business School in Nice, France. I have 12 years of industry experience and over 15 years of academic experience.

Contact me at dominic.okane at edhec.edu.

## Dependencies

FinancePy depends on Numpy, Numba, Scipy and basic python libraries such as os, sys and datetime.

## Changelog

See the changelog for a detailed history of changes.

## Contributions

Contributions are very welcome. There are a number of requirements:

* The code should be Pep8 compliant.
* Comments are required for every class and function and they should be a clear description.
* At least one broad test case and a set of unit tests must be provided for every function.
* Avoid very pythonic constructions. For example a loop is as good as a list comprehension. And with numba it can be faster. Readability is the priority.

## License

 GPL-3.0 License - See the license file in this folder for details.


changes.md
23 August 2023
- Renamed all 'full' price functions to 'dirty' prices
- Added ex_div_days to Bond constructor
- Removed face amount from bond construction
- Face amount for bonds is entered when calculating payments, accrued or principal.
- Face amount is not an intrinsic member of the bond class


examples/TestFinDate.py
###############################################################################
# Copyright (C) 2018, 2019, 2020 Dominic O'Kane
###############################################################################

import numpy as np
import time

import sys
sys.path.append("..")

from financepy.utils.date import Date, dateRange
from financepy.utils.date import FinDateFormatTypes
from financepy.utils.date import setDateFormatType

from FinTestCases import FinTestCases, globalTestCaseMode
testCases = FinTestCases(__file__, globalTestCaseMode)

###############################################################################

setDateFormatType(FinDateFormatTypes.UK_LONGEST)

def test_FinDate():

    start_date = Date(1, 1, 2018)

    assert Date(1, 1, 2018) == Date.fromString('1-1-2018', '%d-%m-%Y')

    testCases.header("DATE", "MONTHS", "CDS DATE")

    for numMonths in range(0, 120):
        nextCDSDate = start_date.nextCDSDate(numMonths)
        testCases.print(str(start_date), numMonths, str(nextCDSDate))

    start_date = Date(1, 1, 2018)

    testCases.header("STARTDATE", "MONTHS", "CDS DATE")

    for numMonths in range(0, 365):
        start_date = start_date.addDays(1)
        nextIMMDate = start_date.nextIMMDate()
        testCases.print(numMonths, str(start_date), str(nextIMMDate))

###############################################################################


def test_FinDateTenors():

    start_date = Date(23, 2, 2018)

    testCases.header("TENOR", "DATE")
    tenor = "5d"
    testCases.print(tenor, start_date.addTenor(tenor))

    tenor = "7D"
    testCases.print(tenor, start_date.addTenor(tenor))

    tenor = "1W"
    testCases.print(tenor, start_date.addTenor(tenor))

    tenor = "4W"
    testCases.print(tenor, start_date.addTenor(tenor))

    tenor = "1M"
    testCases.print(tenor, start_date.addTenor(tenor))

    tenor = "24M"
    testCases.print(tenor, start_date.addTenor(tenor))

    tenor = "2Y"
    testCases.print(tenor, start_date.addTenor(tenor))

    tenor = "10y"
    testCases.print(tenor, start_date.addTenor(tenor))

    tenor = "0m"
    testCases.print(tenor, start_date.addTenor(tenor))

    tenor = "20Y"
    testCases.print(tenor, start_date.addTenor(tenor))

###############################################################################


def test_FinDateRange():

    start_date = Date(1, 1, 2010)

    testCases.header("Tenor", "Dates")

    end_date = start_date.addDays(3)
    tenor = "Default"
    testCases.print(tenor, dateRange(start_date, end_date))

    end_date = start_date.addDays(20)
    tenor = "1W"
    testCases.print(tenor, dateRange(start_date, end_date, tenor))

    tenor = "7D"
    testCases.print(tenor, dateRange(start_date, end_date, tenor))

    testCases.header("Case", "Dates")

    case = "Same start_date"
    testCases.print(case, dateRange(start_date, start_date))
    case = "start_date before end_date"
    testCases.print(case, dateRange(end_date, start_date))

###############################################################################


def test_FinDateAddMonths():

    start_date = Date(1, 1, 2010)

    testCases.header("Months", "Dates")

    months = [1, 3, 6, 9, 12, 24, 36, 48, 60]

    dates = start_date.addMonths(months)

    testCases.header("DATES", "DATE")

    for dt in dates:
        testCases.print("DATE", dt)

###############################################################################


def test_FinDateAddYears():

    start_date = Date(1, 1, 2010)

    testCases.header("Years", "Dates")

    years = [1, 3, 5, 7, 10]
    dates1 = start_date.addYears(years)
    for dt in dates1:
        testCases.print("DATES1", dt)

    years = np.array([1, 3, 5, 7, 10])
    dates2 = start_date.addYears(years)
    for dt in dates2:
        testCases.print("DATES2", dt)

    years = np.array([1.5, 3.25, 5.75, 7.25, 10.0])
    dates3 = start_date.addYears(years)

    for dt in dates3:
        testCases.print("DATES3", dt)

    dt = 1.0/365.0
    years = np.array([1.5+2.0*dt, 3.5-6*dt, 5.75+3*dt, 7.25+dt, 10.0+dt])
    dates4 = start_date.addYears(years)

    for dt in dates4:
        testCases.print("DATES4", dt)

###############################################################################


def test_FinDateSpeed():

    num_steps = 100
    start = time.time()
    dateList = []
    for _ in range(0, num_steps):
        start_date = Date(1, 1, 2010)
        dateList.append(start_date)
    end = time.time()
    elapsed = end - start

    testCases.header("LABEL", "TIME")
    testCases.print("TIMING", elapsed)

    mem = sys.getsizeof(dateList)
    testCases.print("Mem:", mem)


###############################################################################


def test_FinDateFormat():

    dt = Date(20, 10, 2019)
    testCases.header("FORMAT", "DATE")

    for formatType in FinDateFormatTypes:
        setDateFormatType(formatType) 
        testCases.print(formatType.name, dt)

###############################################################################


def test_IntraDay():

    testCases.header("Date1", "Date2", "Diff")
    d1 = Date(20, 10, 2019, 0, 0, 0)
    d2 = Date(25, 10, 2019, 0, 0, 0)
    diff = d2 - d1
    testCases.print(d1, d2, diff)
    testCases.print(d1._excelDate, d2._excelDate, diff)

    ###########################################################################

    d1 = Date(20, 10, 2019, 10, 0, 0)
    d2 = Date(25, 10, 2019, 10, 25, 0)
    diff = d2 - d1
    testCases.print(d1, d2, diff)
    testCases.print(d1._excelDate, d2._excelDate, diff)

    ###########################################################################

    d1 = Date(20, 10, 2019, 10, 0, 0)
    d2 = Date(20, 10, 2019, 10, 25, 30)
    diff = d2 - d1
    testCases.print(d1, d2, diff)
    testCases.print(d1._excelDate, d2._excelDate, diff)

    ###########################################################################

    d1 = Date(19, 10, 2019, 10, 0, 0)
    d2 = Date(20, 10, 2019, 10, 25, 40)
    diff = d2 - d1
    testCases.print(d1, d2, diff)
    testCases.print(d1._excelDate, d2._excelDate, diff)

###############################################################################

def test_FinDateEOM():

    dt = Date(29, 2, 2000)
    assert(dt.isEOM() == True)

    dt = Date(28, 2, 2001)
    assert(dt.isEOM() == True)

    dt = Date(29, 2, 2004)
    assert(dt.isEOM() == True)

    dt = Date(28, 2, 2005)
    assert(dt.isEOM() == True)

    dt = Date(31, 3, 2003)
    assert(dt.isEOM() == True)

    dt = Date(30, 4, 2004)
    assert(dt.isEOM() == True)

    dt = Date(31, 5, 2004)
    assert(dt.isEOM() == True)

    dt = Date(31, 12, 2010)
    assert(dt.isEOM() == True)

    dt = Date(2, 2, 2000)
    assert(dt.EOM().isEOM() == True)

    dt = Date(24, 2, 2001)
    assert(dt.EOM().isEOM() == True)

    dt = Date(22, 2, 2004)
    assert(dt.EOM().isEOM() == True)

    dt = Date(1, 2, 2005)
    assert(dt.EOM().isEOM() == True)

    dt = Date(1, 3, 2003)
    assert(dt.EOM().isEOM() == True)

    dt = Date(3, 4, 2004)
    assert(dt.EOM().isEOM() == True)

    dt = Date(5, 5, 2004)
    assert(dt.EOM().isEOM() == True)

    dt = Date(7, 12, 2010)
    assert(dt.EOM().isEOM() == True)

###############################################################################
    
start = time.time()

test_FinDate()
test_FinDateTenors()
test_FinDateRange()
test_FinDateAddMonths()
test_FinDateAddYears()
test_FinDateSpeed()
test_FinDateFormat()
test_IntraDay()
test_FinDateEOM()

end = time.time()
elapsed = end - start
# print("Elapsed time:", elapsed)

testCases.compareTestCases()

setDateFormatType(FinDateFormatTypes.UK_LONG)
